In [24]:
# Imports section
import numpy as np
import pandas as pd
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import cross_val_score
from sklearn.preprocessing import PolynomialFeatures
from sklearn.metrics import r2_score

## Part 1. Loading the dataset

In [25]:
# Using pandas load the dataset (load remotely, not locally)
data = pd.read_csv('./science_data_large.csv')
# Output the first 15 rows of the data
print(data.head(15))
# Display a summary of the table information (number of datapoints, etc.)
data.info()

    Temperature °C  Mols KCL     Size nm^3
0              469       647  6.244743e+05
1              403       694  5.779610e+05
2              302       975  6.196847e+05
3              779       916  1.460449e+06
4              901        18  4.325726e+04
5              545       637  7.124634e+05
6              660       519  7.006960e+05
7              143       869  2.718260e+05
8               89       461  8.919803e+04
9              294       776  4.770210e+05
10             991       117  2.441771e+05
11             307       781  5.006455e+05
12             206        70  3.145200e+04
13             437       599  5.390215e+05
14             566        75  9.185271e+04
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Temperature °C  1000 non-null   int64  
 1   Mols KCL        1000 non-null   int64  
 2   Size nm^3       1000 n

## Part 2. Splitting the dataset

In [26]:
# Take the pandas dataset and split it into our features (X) and label (y)
X = data[['Temperature °C', 'Mols KCL']]
y = data['Size nm^3']

# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.1, random_state = 50)
print(X_train.shape)
print(y_train.shape)

(900, 2)
(900,)


## Part 3. Perform a Linear Regression

In [27]:
# Use sklearn to train a model on the training set
model = LinearRegression()
model.fit(X_train, y_train)
# Create a sample datapoint and predict the output of that sample with the trained model
sample = [[25, 0.5]]
predicted_size = model.predict(sample)
# Report on the score for that model, in your own words (markdown, not code) explain what the score means
score = model.score(X_test, y_test)
print(f'Score: {score: .3f}')
# Extract the coefficents and intercept from the model and write an equation for your h(x) using LaTeX
coefficents = model.coef_
intercept = model.intercept_
print(f'coefficents: {coefficents}')
print(f'intercept: {intercept}')

Score:  0.844
coefficents: [ 871.41449757 1038.23844296]
intercept: -416367.37884415605


/home/codespace/.local/lib/python3.12/site-packages/sklearn/base.py:493: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  warnings.warn(


Score: The score represents the R^2 score to evluate the accuracy for the test. R^2 is a statistical measure to represent the proportion of the variance for variables. Most of the time the score is in range [0, 1]. If the score is close to 1, it indicates that the model performs well on the test data; if the score is close to 0, it means that the model is unable to explain the variation in the target variable. 

Equation: $y = -416367.38 + 871.41x_1 + 1038.24x_2$

## Part 4. Use Cross Validation

In [28]:
# Use the cross_val_score function to repeat your experiment across many shuffles of the data

score = cross_val_score(model, X, y, cv = 10, scoring = 'r2')
print(f'Score: {score}')
avg_score = np.mean(score)
print(f'Average score: {avg_score}')
# Report on their finding and their significance


Score: [0.81123596 0.86440978 0.87808742 0.86561069 0.87495621 0.84484397
 0.87941022 0.86349411 0.78353682 0.88686516]
Average score: 0.8552450341984701


Finding: The average score using cross validation is larger than the score calculated in part 3. This indicates that the model performs well in multiple different data splits, which means that it has a good stability and generalization. 

Significance: The cross validation helps to assess the liability of the model using multiple different sets of the data. Also, the cross validation can use the full dataset since every sample is used in training and testing. 


## Part 5. Using Polynomial Regression

In [29]:
# Using the PolynomialFeatures library perform another regression on an augmented dataset of degree 2

# Report on the metrics and output the resultant equation as you did in Part 3.
poly = PolynomialFeatures(degree=2)
X_train_poly = poly.fit_transform(X_train)
X_test_poly = poly.transform(X_test)
poly_model = LinearRegression()
poly_model.fit(X_train_poly, y_train)
#predict
y_pred = poly_model.predict(X_test_poly)
#score
score = r2_score(y_test, y_pred)
print(f'Score: {score}')
coefficients = poly_model.coef_
intercept = poly_model.intercept_
formatted_coefficients = [f'{coef:.3f}' for coef in coefficients]
print(f'Coefficients: {formatted_coefficients}')
print(f'Intercept: {intercept:3f}')

Score: 1.0
Coefficients: ['0.000', '12.000', '-0.000', '-0.000', '2.000', '0.029']
Intercept: 0.000010


Equation: $y = 0.000010 + 12.000 \cdot x_1 + 2.000 \cdot (x_1 \times x_2) + 0.029 \cdot x_2^2$